mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp

```diff
@@ -173,14 +173,62 @@ class SelectOpConversion : public OpConversionPattern<arith::SelectOp> {
   }
 };
 
+// Floating-point to integer, integer to floating-point conversions.
+template <typename CastOp, bool opHasTruncateSemantics>
+class CastOpConversion : public OpConversionPattern<CastOp> {
+private:
+  bool floatToIntTruncates;
+
+public:
+  CastOpConversion(const TypeConverter &typeConverter, MLIRContext *context,
+                   bool optionFloatToIntTruncates)
+      : OpConversionPattern<CastOp>(typeConverter, context),
+        floatToIntTruncates(optionFloatToIntTruncates) {}
+
+  LogicalResult
+  matchAndRewrite(CastOp castOp, typename CastOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Type operandType = adaptor.getIn().getType();
+    if (!operandType.isIntOrFloat())
+      return rewriter.notifyMatchFailure(castOp,
+                                         "casts are only supported on scalars");
+
+    if (!(emitc::isSupportedFloatType(operandType) ||
+          emitc::isSupportedIntegerType(operandType)))
+      return rewriter.notifyMatchFailure(castOp,
+                                         "unsupported cast source type");
+
+    if (opHasTruncateSemantics && !floatToIntTruncates)
+      return rewriter.notifyMatchFailure(
+          castOp,
+          "conversion requires casts to use truncation as rounding mode");
```

Comment:
```suggestion
          "conversion currently only supports cast conversion if the EmitC cast can be assumed to be truncating");
```

Reply:
I suggest 
```
"conversion currently requires casts to use truncation as rounding mode"
```
instead. 

I don't mention EmitC in the comment as the hypothesis isn't on EmitC but on the downstream compiler that must preserve these semantics. EmitC has no knowledge of the rounding mode.

Reply:
Sounds good